webapp/channels/src/utils/desktop_api.ts

```diff
@@ -0,0 +1,288 @@
+            listener(message);
+        });
+    };
+
+    /**
+     * @deprecated
+     */
+    addPostMessageListener = (channel: string, listener: (message: any) => void) => {
+        if (this.postMessageListeners?.has(channel)) {
+            this.postMessageListeners.set(channel, this.postMessageListeners.get(channel)!.add(listener));
+        } else {
+            this.postMessageListeners?.set(channel, new Set([listener]));
+        }
+    };
+
+    /**
+     * @deprecated
+     */
+    removePostMessageListener = (channel: string, listener: (message: any) => void) => {
+        const set = this.postMessageListeners?.get(channel);
+        set?.delete(listener);
+        if (set?.size) {
+            this.postMessageListeners?.set(channel, set);
+        } else {
+            this.postMessageListeners?.delete(channel);
+        }
+    };
+
+    /**
+     * @deprecated
+     */
+    invokeWithMessaging = <T, T2>(
+        sendChannel: string,
+        sendData?: T,
+        receiveChannel?: string,
+    ) => {
+        return new Promise<T2>((resolve) => {
+            /* create and register a temporary listener if necessary */
+            const response = ({origin, data: {type, message}}: {origin: string; data: {type: string; message: T2}}) => {
+                /* ignore messages from other frames */
+                if (origin !== window.location.origin) {
+                    return;
+                }
+
+                /* ignore messages from other channels */
+                if (type !== (receiveChannel ?? sendChannel)) {
+                    return;
+                }
+
+                /* clean up listener and resolve */
+                window.removeEventListener('message', response);
+                resolve(message);
+            };
+
+            window.addEventListener('message', response);
+            window.postMessage(
+                {type: sendChannel, message: sendData},
+                window.location.origin,
+            );
+        });
+    };
+}
+
+const desktopApp = new DesktopApp();
+export default desktopApp;
+
+/*******************************
+ * API Methods
+ *******************************/
+
+/**
+ * Invokes
+ */
+
+export const getBrowserHistoryStatus = async () => {
+    if (window.desktopAPI?.requestBrowserHistoryStatus) {
+        return window.desktopAPI.requestBrowserHistoryStatus();
+    }
+
+    const {enableBack, enableForward} = await desktopApp.invokeWithMessaging<void, {enableBack: boolean; enableForward: boolean}>(
+        'history-button',
+        undefined,
+        'history-button-return',
+    );
+
+    return {
+        canGoBack: enableBack,
+        canGoForward: enableForward,
+    };
+};
+
+/**
+ * Listeners
+ */
+
+export const onUserActivityUpdate = (listener: (userIsActive: boolean, idleTime: number, isSystemEvent: boolean) => void) => {
+    if (window.desktopAPI?.onUserActivityUpdate) {
+        window.desktopAPI.onUserActivityUpdate(listener);
+        return () => {};
```

Comment:
Does `desktopAPI.onUserActivityUpdate` not return a cleanup function of its own?

I know we discussed making the desktop app responsible for final cleanup for the app when it's closed or refresh, but these are going to leak memory if the component that registers these callbacks is ever unmounted. Even components like `LoggedIn` and `ChannelController` are remounted when moving to areas like the System Console, so I think we need cleanup functions for the newer versions of these

Reply:
Yeah I mentioned this in the other PR, we need to return the `unregister` call, but unfortunately that will break all listeners on the same channel, which isn't ideal.

I did want to avoid making `on`/`off` for everything, but that might be the only way.

Reply:
I don't suppose there's a way for the desktop API to also return an `off` when you call an `on`? If this could just `return window.desktopAPI.onUserActivityUpdate(listener)`, that seems ideal

Reply:
That would require changes to the Desktop App side of things, and is another thing to maintain for every single listener call. Let me see if I can find an intelligent way to wrap it and if I can, we'll do that since that would obviously be better for the Web App side of things.

If not, I'll come back here and we'll figure out something else.